bpf/flows.c

```diff
@@ -19,43 +47,64 @@
 #define INGRESS 0
 #define EGRESS 1
 
-// TODO: for performance reasons, replace the ring buffer by a hashmap and
-// aggregate the flows here instead of the Go Accounter
+// Common Ringbuffer as a conduit for ingress/egress maps to userspace
 struct {
     __uint(type, BPF_MAP_TYPE_RINGBUF);
     __uint(max_entries, 1 << 24);
 } flows SEC(".maps");
 
+struct {
+    __uint(type, BPF_MAP_TYPE_PERCPU_HASH);
+    __type(key, flow_id_v);
+    __type(value, flow_metrics);
+    __uint(max_entries, INGRESS_MAX_ENTRIES);
+} xflow_metric_map_ingress SEC(".maps");
+
+struct {
+    __uint(type, BPF_MAP_TYPE_PERCPU_HASH);
+    __type(key, flow_id_v);
+    __type(value, flow_metrics);
+    __uint(max_entries, EGRESS_MAX_ENTRIES);
+} xflow_metric_map_egress SEC(".maps");
+
 // Constant definitions, to be overridden by the invoker
 volatile const u32 sampling = 0;
 
 const u8 ip4in6[] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0xff, 0xff};
 
 // sets flow fields from IPv4 header information
-static inline int fill_iphdr(struct iphdr *ip, void *data_end, struct flow *flow) {
+// Flags is highlight any protocol specific info
+static inline int fill_iphdr(struct iphdr *ip, void *data_end, flow_id_v *flow_id, u32 *flags) {
     if ((void *)ip + sizeof(*ip) > data_end) {
         return DISCARD;
     }
 
-    __builtin_memcpy(flow->network.src_ip.s6_addr, ip4in6, sizeof(ip4in6));
-    __builtin_memcpy(flow->network.dst_ip.s6_addr, ip4in6, sizeof(ip4in6));
-    __builtin_memcpy(flow->network.src_ip.s6_addr + sizeof(ip4in6), &ip->saddr, sizeof(ip->saddr));
-    __builtin_memcpy(flow->network.dst_ip.s6_addr + sizeof(ip4in6), &ip->daddr, sizeof(ip->daddr));
-    flow->transport.protocol = ip->protocol;
-
+    __builtin_memcpy(flow_id->src_ip.s6_addr, ip4in6, sizeof(ip4in6));
+    __builtin_memcpy(flow_id->dst_ip.s6_addr, ip4in6, sizeof(ip4in6));
+    __builtin_memcpy(flow_id->src_ip.s6_addr + sizeof(ip4in6), &ip->saddr, sizeof(ip->saddr));
+    __builtin_memcpy(flow_id->dst_ip.s6_addr + sizeof(ip4in6), &ip->daddr, sizeof(ip->daddr));
+    flow_id->protocol = ip->protocol;
+    flow_id->src_port = 0;
+    flow_id->dst_port = 0;
     switch (ip->protocol) {
     case IPPROTO_TCP: {
         struct tcphdr *tcp = (void *)ip + sizeof(*ip);
         if ((void *)tcp + sizeof(*tcp) <= data_end) {
-            flow->transport.src_port = __bpf_ntohs(tcp->source);
-            flow->transport.dst_port = __bpf_ntohs(tcp->dest);
+            flow_id->src_port = __bpf_ntohs(tcp->source);
+            flow_id->dst_port = __bpf_ntohs(tcp->dest);
+            if (tcp->fin) {
+                *flags= *flags | TCP_FIN_FLAG;
+            }
+            if (tcp->rst) {
+                *flags= *flags | TCP_RST_FLAG;
+            }
```

Comment:
should we add more than fin/rst? SYN maybe ?

Reply:
Ideally, I guess that information could be inferred by FlowStartTime variable. But fundamentally, yes,  We can add more flags based on the use-case.

Reply:
We will definitely need to discuss more about which flags could be captured.
it could be added as additional enhancements on top?